### PR TITLE
Create functionalities.py

### PR DESCRIPTION
**Pull request by AkshIyer-AI**
Summary of Changes for Pull Request

- Introduced a DataPipeline class to manage a simple ETL (Extract, Transform, Load) workflow using SQLite as the backend database.
- Implemented methods for:
  - Initializing the database and creating a users table if it does not exist.
  - Loading user data from a CSV file.
  - Inserting user data into the database.
  - Fetching all user records from the database.
  - Exporting user data to a JSON file.
- Added a run_pipeline function to orchestrate the pipeline steps: database initialization, CSV loading, data insertion, data retrieval, and JSON export.
- Utilized standard Python libraries: csv, json, sqlite3, and pandas (though pandas is imported but not used in this version).
- No external dependencies beyond the Python standard library.
- The script can be executed directly to run the full pipeline on a users.csv file.

### CODE REVIEW

File: code_for_review.py [hunk 1]
```diff
@@ -1,4 +1,104 @@
...
+from typing import List, Dict, Any, Optional
+
+
+def read_json_file(filepath: str) -> Dict[str, Any]:
+    """Read JSON from a file and return a dictionary."""
+    if not os.path.exists(filepath):
+        raise FileNotFoundError(f"File not found: {filepath}")
+    with open(filepath, "r", encoding="utf-8") as f:
+        return json.load(f)
+
```
Comment (AkshIyer-AI) [inline code_for_review.py:15]: Avoid exposing full file paths in error messages. Use generic messages for file errors.

File: code_for_review.py [hunk 2]
```diff
@@ -1,4 +1,104 @@
...
+    if not numbers:
+        raise ValueError("List of numbers cannot be empty")
+    return {
+        "mean": statistics.mean(numbers),
+        "median": statistics.median(numbers),
+        "stdev": statistics.stdev(numbers) if len(numbers) > 1 else 0.0,
+        "min": min(numbers),
+        "max": max(numbers),
+    }
+
```
Comment (AkshIyer-AI) [inline code_for_review.py:34]: Ensure error messages for invalid statistics input are generic if shown to users; avoid leaking sensitive data.

File: code_for_review.py [hunk 3]
```diff
@@ -1,4 +1,104 @@
...
+    return 1 if n <= 1 else n * factorial(n - 1)
+
+
+def fetch_github_user(username: str) -> Optional[Dict[str, Any]]:
+    """Fetch a GitHub user's public info using GitHub API."""
+    url = f"https://api.github.com/users/{username}"
+    response = requests.get(url, timeout=10)
+    if response.status_code == 200:
+        return response.json()
+    return None
```
Comment (AkshIyer-AI) [inline code_for_review.py:49]: Wrap the requests.get call in try/except to handle HTTP/network errors gracefully. Validate 'username' before using it in the URL.

File: code_for_review.py [hunk 4]
```diff
@@ -1,4 +1,104 @@
...
+        if not all(isinstance(n, (int, float)) for n in numbers):
+            raise TypeError("All elements must be int or float")
+        self.data = numbers
+
+    def normalize_data(self) -> List[float]:
+        """Normalize data between 0 and 1."""
+        if not self.data:
+            raise ValueError("No data loaded")
+        min_val, max_val = min(self.data), max(self.data)
+        if min_val == max_val:
```
Comment (AkshIyer-AI) [inline code_for_review.py:70]: Check that all loaded values are finite numbers (not NaN or inf) before processing.

File: code_for_review.py [hunk 5]
```diff
@@ -1,4 +1,104 @@
...
+            return [0.5 for _ in self.data]
+        return [(x - min_val) / (max_val - min_val) for x in self.data]
+
+    def get_stats(self) -> Dict[str, float]:
+        """Return statistics of the loaded data."""
+        return calculate_statistics(self.data)
+
```
Comment (AkshIyer-AI) [inline code_for_review.py:77]: Handle errors when no data is loaded for normalization.

File: code_for_review.py [hunk 6]
```diff
@@ -1,4 +1,104 @@
...
+        write_json_file(filepath, {"name": self.name, "data": self.data})
+
+
+# ------------------- Example usage -------------------
+if __name__ == "__main__":
+    # Factorial demo
+    print("Factorial of 5:", factorial(5))
+
+    # GitHub API demo
+    user_info = fetch_github_user("torvalds")
```
Comment (AkshIyer-AI) [inline code_for_review.py:89]: Add try/except to handle errors from 'write_json_file' in 'save_to_file'.

File: code_for_review.py [hunk 7]
```diff
@@ -1,4 +1,104 @@
...
+    processor.load_data([10, 20, 30, 40, 50])
+    print("Normalized data:", processor.normalize_data())
+    print("Statistics:", processor.get_stats())
+
+    # Save to file
+    processor.save_to_file("output.json")
+    print("Data saved to output.json")
 
 class usermanager:
     def __init__(self, NAME, Age):
```
Comment (AkshIyer-AI) [inline code_for_review.py:104]: Do not print or log sensitive user information from API responses in production.

File: code_for_review.py [hunk 8]
```diff
@@ -1,4 +1,104 @@
...
+        """Normalize data between 0 and 1."""
+        if not self.data:
+            raise ValueError("No data loaded")
+        min_val, max_val = min(self.data), max(self.data)
+        if min_val == max_val:
+            return [0.5 for _ in self.data]
+        return [(x - min_val) / (max_val - min_val) for x in self.data]
+
+    def get_stats(self) -> Dict[str, float]:
+        """Return statistics of the loaded data."""
```
Comment (AkshIyer-AI) [inline code_for_review.py:75]: Parameter and instance variable names should use lowercase with underscores (e.g., 'name', 'age') for consistency.

File: code_for_review.py [hunk 9]
```diff
@@ -1,4 +1,104 @@
...
+        return calculate_statistics(self.data)
+
+    def save_to_file(self, filepath: str) -> None:
```
Comment (AkshIyer-AI) [inline code_for_review.py:78]: Rename methods 'getname' and 'getage' to 'get_name' and 'get_age' (snake_case).

File: code_for_review.py [hunk 10]
```diff
@@ -1,4 +1,104 @@
...
+        """Save the current data to a JSON file."""
+        write_json_file(filepath, {"name": self.name, "data": self.data})
+
+
+# ------------------- Example usage -------------------
```
Comment (AkshIyer-AI) [inline code_for_review.py:83]: Rename 'CalculateAgeInMonths' to 'calculate_age_in_months', and fix indentation.

File: code_for_review.py [hunk 11]
```diff
@@ -1,4 +1,104 @@
...
+    # Factorial demo
+    print("Factorial of 5:", factorial(5))
+
+    # GitHub API demo
+    user_info = fetch_github_user("torvalds")
+    if user_info:
+        print("GitHub User:", user_info["login"], "| Public repos:", user_info["public_repos"])
+
+    # Data processing demo
+    processor = DataProcessor("SampleData")
```
Comment (AkshIyer-AI) [inline code_for_review.py:94]: Rename 'misSpelled_functn' to a meaningful, correctly spelled snake_case name (e.g., 'double_value').

File: code_for_review.py [hunk 12]
```diff
@@ -1,4 +1,104 @@
...
+    if response.status_code == 200:
+        return response.json()
+    return None
+
+
+class DataProcessor:
+    """A class to handle data processing tasks."""
+
+    def __init__(self, name: str):
+        self.name = name
```
Comment (AkshIyer-AI) [inline code_for_review.py:56]: Sanitize and validate 'username' input before using in the GitHub API URL to prevent injection risks.

File: code_for_review.py [hunk 13]
```diff
@@ -33,3 +133,4 @@ def Main():
     print(misSpelled_functn(5))
```
Comment (AkshIyer-AI) [inline code_for_review.py:134]: 'CalculateAgeInMonths' should validate input is a non-negative int or float to prevent logic errors.

File: code_for_review.py [hunk 14]
```diff
@@ -1,4 +1,104 @@
...
+
+def read_json_file(filepath: str) -> Dict[str, Any]:
+    """Read JSON from a file and return a dictionary."""
+    if not os.path.exists(filepath):
+        raise FileNotFoundError(f"File not found: {filepath}")
+    with open(filepath, "r", encoding="utf-8") as f:
+        return json.load(f)
+
+
+def write_json_file(filepath: str, data: Dict[str, Any]) -> None:
```
Comment (AkshIyer-AI) [inline code_for_review.py:17]: Avoid exposing full file paths in error messages to prevent information leakage. Use generic error messages in FileNotFoundError.

File: code_for_review.py [hunk 15]
```diff
@@ -1,4 +1,104 @@
...
+    """Write a dictionary as JSON into a file."""
```
Comment (AkshIyer-AI) [inline code_for_review.py:18]: Wrap JSON file reading and parsing in try/except to handle I/O and decoding errors gracefully.

File: code_for_review.py [hunk 16]
```diff
@@ -1,4 +1,104 @@
...
+    with open(filepath, "w", encoding="utf-8") as f:
+        json.dump(data, f, indent=2)
+
+
+def calculate_statistics(numbers: List[float]) -> Dict[str, float]:
+    """Return basic statistics for a list of numbers."""
```
Comment (AkshIyer-AI) [inline code_for_review.py:24]: Wrap JSON file writing in try/except to handle possible I/O or serialization errors.

File: code_for_review.py [hunk 17]
```diff
@@ -1,4 +1,104 @@
...
+        raise ValueError("Factorial not defined for negative numbers")
+    return 1 if n <= 1 else n * factorial(n - 1)
+
+
+def fetch_github_user(username: str) -> Optional[Dict[str, Any]]:
+    """Fetch a GitHub user's public info using GitHub API."""
+    url = f"https://api.github.com/users/{username}"
+    response = requests.get(url, timeout=10)
+    if response.status_code == 200:
+        return response.json()
```
Comment (AkshIyer-AI) [inline code_for_review.py:48]: Limit factorial input size or use an iterative approach to prevent RecursionError or DoS with large values.

File: code_for_review.py [hunk 18]
```diff
@@ -1,4 +1,104 @@
...
+    return None
+
+
+class DataProcessor:
+    """A class to handle data processing tasks."""
+
```
Comment (AkshIyer-AI) [inline code_for_review.py:54]: Wrap requests.get in try/except to handle network errors. Validate 'username' parameter before using in URL.

File: code_for_review.py [hunk 19]
```diff
@@ -1,4 +1,104 @@
...
+    def __init__(self, name: str):
```
Comment (AkshIyer-AI) [inline code_for_review.py:55]: Handle non-200 API responses and log errors safely without exposing sensitive information.

File: code_for_review.py [hunk 20]
```diff
@@ -1,4 +1,104 @@
...
+            return [0.5 for _ in self.data]
+        return [(x - min_val) / (max_val - min_val) for x in self.data]
+
+    def get_stats(self) -> Dict[str, float]:
+        """Return statistics of the loaded data."""
+        return calculate_statistics(self.data)
+
+    def save_to_file(self, filepath: str) -> None:
+        """Save the current data to a JSON file."""
+        write_json_file(filepath, {"name": self.name, "data": self.data})
```
Comment (AkshIyer-AI) [inline code_for_review.py:80]: Wrap write_json_file in try/except in save_to_file to handle I/O errors.

File: code_for_review.py [hunk 21]
```diff
@@ -1,4 +1,104 @@
...
+# ------------------- Example usage -------------------
+if __name__ == "__main__":
+    # Factorial demo
+    print("Factorial of 5:", factorial(5))
+
+    # GitHub API demo
+    user_info = fetch_github_user("torvalds")
+    if user_info:
+        print("GitHub User:", user_info["login"], "| Public repos:", user_info["public_repos"])
+
```
Comment (AkshIyer-AI) [inline code_for_review.py:92]: Avoid printing sensitive user data directly. Sanitize output if this will be adapted for real user data.

File: code_for_review.py [hunk 22]
```diff
@@ -1,4 +1,104 @@
-import os,sys
+import os
```
Comment (AkshIyer-AI) [inline code_for_review.py:1]: Unused imports: Remove unused imports (`math`, `sys`) to reduce attack surface and improve maintainability. If `sys` and `math` are not used elsewhere, removing them is a good cleanup.

File: code_for_review.py [hunk 23]
```diff
@@ -1,4 +1,104 @@
-import os,sys
+import os
+import json
```
Comment (AkshIyer-AI) [inline code_for_review.py:2]: Remove unused import: 'math' is imported but not used. Keeping unused imports reduces code readability and clarity.

File: code_for_review.py [hunk 24]
```diff
@@ -1,4 +1,104 @@
-import os,sys
+import os
+import json
+import math
+import statistics
+import requests
```
Comment (AkshIyer-AI) [inline code_for_review.py:5]: Dependency security: Ensure that the 'requests' library is version pinned in your requirements.txt or equivalent for security. In addition, wrap external calls (see line 46) in try/except to handle errors robustly. 'requests' verifies SSL by default – do not disable SSL verification.

File: code_for_review.py [hunk 25]
```diff
@@ -1,4 +1,104 @@
-import os,sys
+import os
+import json
+import math
+import statistics
+import requests
+from typing import List, Dict, Any, Optional
+
+
+def read_json_file(filepath: str) -> Dict[str, Any]:
```
Comment (AkshIyer-AI) [inline code_for_review.py:9]: Docstring for 'read_json_file': Consider explaining why strict error handling (e.g., for missing files, JSON decode errors) is chosen, not just what the function does.

File: code_for_review.py [hunk 26]
```diff
@@ -1,4 +1,104 @@
...
+import json
+import math
+import statistics
+import requests
+from typing import List, Dict, Any, Optional
+
+
+def read_json_file(filepath: str) -> Dict[str, Any]:
+    """Read JSON from a file and return a dictionary."""
+    if not os.path.exists(filepath):
```
Comment (AkshIyer-AI) [inline code_for_review.py:11]: Robustness: File operations lack error handling. Wrap JSON file reads in try/except to properly handle IO errors, missing files, and JSON decode errors, and ensure error messages do not leak file paths or sensitive info.

File: code_for_review.py (hunk 14 above)
Comment (AkshIyer-AI) [inline code_for_review.py:17]: Docstring for 'write_json_file': Consider expanding to explain the implementation decisions, e.g., why you use UTF-8 encoding and formatted output.

File: code_for_review.py [hunk 27]
```diff
@@ -1,4 +1,104 @@
...
+        raise FileNotFoundError(f"File not found: {filepath}")
+    with open(filepath, "r", encoding="utf-8") as f:
+        return json.load(f)
+
+
+def write_json_file(filepath: str, data: Dict[str, Any]) -> None:
+    """Write a dictionary as JSON into a file."""
+    with open(filepath, "w", encoding="utf-8") as f:
```
Comment (AkshIyer-AI) [inline code_for_review.py:19]: Robustness: File write operations should use try/except blocks to catch IO and permission errors, avoiding leakage of file paths or sensitive information in error messages.

File: code_for_review.py [hunk 28]
```diff
@@ -1,4 +1,104 @@
...
+        json.dump(data, f, indent=2)
+
+
+def calculate_statistics(numbers: List[float]) -> Dict[str, float]:
```
Comment (AkshIyer-AI) [inline code_for_review.py:23]: Docstring for 'calculate_statistics': Suggest expanding to cover how edge cases and single-value lists are handled (e.g., stdev returns 0.0).

File: code_for_review.py [hunk 29]
```diff
@@ -1,4 +1,104 @@
...
+    return {
+        "mean": statistics.mean(numbers),
+        "median": statistics.median(numbers),
+        "stdev": statistics.stdev(numbers) if len(numbers) > 1 else 0.0,
+        "min": min(numbers),
+        "max": max(numbers),
+    }
+
+
+def factorial(n: int) -> int:
```
Comment (AkshIyer-AI) [inline code_for_review.py:36]: Factorial implementation: The recursive factorial function will hit Python's recursion limit for large n; consider using an iterative approach for better scalability and explain this edge case in the docstring.

File: code_for_review.py [hunk 30]
```diff
@@ -1,4 +1,104 @@
...
+    """Compute factorial of n recursively."""
+    if n < 0:
+        raise ValueError("Factorial not defined for negative numbers")
+    return 1 if n <= 1 else n * factorial(n - 1)
+
+
+def fetch_github_user(username: str) -> Optional[Dict[str, Any]]:
```
Comment (AkshIyer-AI) [inline code_for_review.py:43]: Input validation and error handling for fetch_github_user: Validate the 'username' parameter for valid GitHub username formats to prevent malformed requests or injection. Also, fetch_github_user does not handle exceptions for network errors or bad responses; wrap the requests.get call in try/except and clarify expected behavior on errors in the docstring.

File: code_for_review.py [hunk 31]
```diff
@@ -1,4 +1,104 @@
...
+    """Fetch a GitHub user's public info using GitHub API."""
+    url = f"https://api.github.com/users/{username}"
+    response = requests.get(url, timeout=10)
```
Comment (AkshIyer-AI) [inline code_for_review.py:46]: API call security and error handling: External API calls need try/except wrappers to handle network failures, timeouts, or bad responses. When using requests, do not disable SSL certificate verification. If requests.get(url, verify=False) is ever used, it will be insecure.

File: code_for_review.py [hunk 32]
```diff
@@ -1,4 +1,104 @@
...
+    if response.status_code == 200:
+        return response.json()
+    return None
+
+
+class DataProcessor:
```
Comment (AkshIyer-AI) [inline code_for_review.py:52]: Class docstring for 'DataProcessor': Expand the class-level docstring to clarify use cases, the meaning of 'name' and 'data', and behavior in edge cases.

File: code_for_review.py [hunk 33]
```diff
@@ -1,4 +1,104 @@
...
+    """A class to handle data processing tasks."""
+
+    def __init__(self, name: str):
+        self.name = name
+        self.data: List[float] = []
+
+    def load_data(self, numbers: List[float]) -> None:
+        """Load numeric data into the processor."""
+        if not all(isinstance(n, (int, float)) for n in numbers):
```
Comment (AkshIyer-AI) [inline code_for_review.py:61]: Type checks with booleans: isinstance(n, (int, float)) will accept bools (since bool is a subclass of int). If booleans are not intended as data points, add 'and not isinstance(n, bool)' to exclude them.

File: code_for_review.py [hunk 34]
```diff
@@ -1,4 +1,104 @@
...
+            raise TypeError("All elements must be int or float")
+        self.data = numbers
+
+    def normalize_data(self) -> List[float]:
```
Comment (AkshIyer-AI) [inline code_for_review.py:65]: Clarify docstring for 'normalize_data': Explain the rationale for returning [0.5] for all values when min == max (e.g., to avoid division by zero and keep a normalized output).

File: code_for_review.py [hunk 35]
```diff
@@ -1,4 +1,104 @@
...
+        """Normalize data between 0 and 1."""
+        if not self.data:
+            raise ValueError("No data loaded")
+        min_val, max_val = min(self.data), max(self.data)
+        if min_val == max_val:
+            return [0.5 for _ in self.data]
+        return [(x - min_val) / (max_val - min_val) for x in self.data]
+
+    def get_stats(self) -> Dict[str, float]:
```
Comment (AkshIyer-AI) [inline code_for_review.py:74]: Docstring for 'get_stats': Expand to clarify the structure of the returned statistics and why calculate_statistics is reused.

File: code_for_review.py (hunk 9 above)
Comment (AkshIyer-AI) [inline code_for_review.py:78]: Docstring for 'save_to_file': Suggest expanding to describe the file format, what is being exported (normalized data, name, etc.), and why this method is present.

File: code_for_review.py (hunk 20 above)
Comment (AkshIyer-AI) [inline code_for_review.py:80]: Error handling: File write operation in save_to_file should be wrapped in a try/except block to handle IO problems. Do not leak sensitive information if errors occur.

File: code_for_review.py (hunk 10 above)
Comment (AkshIyer-AI) [inline code_for_review.py:83]: Testing: There is example usage, but no formal tests. Consider adding a test suite (using e.g. pytest or unittest) to check edge cases (empty data, bad input, API/network failures, etc.) and ensure reliability.

File: code_for_review.py [hunk 36]
```diff
@@ -1,4 +1,104 @@
...
+
+# ------------------- Example usage -------------------
+if __name__ == "__main__":
+    # Factorial demo
+    print("Factorial of 5:", factorial(5))
+
+    # GitHub API demo
+    user_info = fetch_github_user("torvalds")
+    if user_info:
+        print("GitHub User:", user_info["login"], "| Public repos:", user_info["public_repos"])
```
Comment (AkshIyer-AI) [inline code_for_review.py:91]: Robustness: When accessing user_info fields (e.g., user_info['login'], user_info['public_repos']), use dict.get() or add KeyError handling for cases where these keys may be absent in the API response. If printing to web output, escape fields to prevent injection/XSS.

File: code_for_review.py [hunk 37]
```diff
@@ -1,4 +1,104 @@
...
+
+    # Data processing demo
+    processor = DataProcessor("SampleData")
+    processor.load_data([10, 20, 30, 40, 50])
+    print("Normalized data:", processor.normalize_data())
+    print("Statistics:", processor.get_stats())
+
+    # Save to file
+    processor.save_to_file("output.json")
```
Comment (AkshIyer-AI) [inline code_for_review.py:100]: Output file location: 'output.json' is written in the current working directory. Consider allowing the output path to be configurable, and robustly handle file-system errors (permissions, disk full, etc.).
